V2.b

brennan-obc opened the pull request:
#crawler.ts commits
- added 'MatchSummary', an interface for the properties extracted from the GET request HTML
- renamed 'getMatchIds' to 'getMatchSummaries', returns a list of 'MatchSummary' objects (including but not limited to match ID)
- added 2 helper functions:
  - fetchMatchHistoryHTML(): fetches match history w/ GET request, returns HTML as string
  - extractMatchSummaries(): parses & formats HTML w/ Cheerio, creates array of MatchSummary objects 
  - getMatchData(): fetches match data w/ POST call containing match ID for each match, using a discrete user-agent header--returns JSON object with properties that almost align w/ 'CharacterDetails' 
  - added regular expressions to format 'getMatchData()' JSON objects to better fit 'CharacterDetails', and to remove unwanted HTML
  - added Bottleneck to control concurrency while fetching match data
- created 'getMatchDetails():
  - restricts maximum concurrent operations to 32
  - fetches match history (raw MatchSummary), extracts match summaries (MatchSummary), and gets match data (CharacterDetail)
  - merges 'MatchSummary' and 'CharacterDetail' based on match ID, creates array of 'MatchDetails' objects

#crawler.test.ts commits
- created new matching 'test.html' and 'test.json' files using GET and POST calls on Dumpster @ Blackrock
- updated test on 'getMatchIds' to test 'getMatchSummaries', updated length and matchId expectations to fit new test documents
- created test "WarmaneCrawler does not exceed 32 concurrent operations"
- created test "The crawler can return a user-readable & database-ready list of match summaries and character details when provided an HTML document and JSON data"
  ! note: this test will be more finely tuned and beefed up once the database component is underway

#test.html & test.json
- created new test files based on a GET (test.html) and a POST (test.json) done on Dumpster/BlackRock

#package.json
- added Bottleneck
- added Parse5 (didn't end up using, at least not yet)

brennan-obc committed:
- renamed 'getMatchIds' to 'getMatchSummaries', returns a list of 'MatchSummary' objects (including but not limited to match ID)
- added 2 helper functions:
  - fetchMatchHistoryHTML(): fetches match history w/ GET request, returns HTML as string
  - extractMatchSummaries(): parses & formats HTML w/ Cheerio, creates array of MatchSummary objects
  - getMatchData(): fetches match data w/ POST call containing match ID for each match, using a discrete user-agent header--returns JSON object with properties that almost align w/ 'CharacterDetails'
  - added regular expressions to format 'getMatchData()' JSON objects to better fit 'CharacterDetails', and to remove unwanted HTML
  - added Bottleneck to control concurrency while fetching match data
- created 'getMatchDetails():
  - restricts maximum concurrent operations to 32
  - fetches match history (raw MatchSummary), extracts match summaries (MatchSummary), and gets match data (CharacterDetail)
  - merges 'MatchSummary' and 'CharacterDetail' based on match ID, creates array of 'MatchDetails' objects

brennan-obc committed:
- created new matching 'test.html' and 'test.json' files using GET and POST calls on Dumpster @ Blackrock
- updated test on 'getMatchIds' to test 'getMatchSummaries', updated length and matchId expectations to fit new test documents
- created test "WarmaneCrawler does not exceed 32 concurrent operations"
- created test "The crawler can return a user-readable & database-ready list of match summaries and character details when provided an HTML document and JSON data"
  ! note: this test will be more finely tuned and beefed up once the database component is underway

brennan-obc committed:
- updated test files to current match data in order for HTML to match JSON

brennan-obc committed:
- added Bottleneck
- added Parse5 (unused)

brennan-obc committed:
- added Bottleneck
- added Parse5 (unused)

briansimoni commented on src/lib/crawler/crawler.ts:
```diff
@@ -1,68 +1,222 @@
+          character_details: characterDetails,
+        };
+
+        console.log(matchDetailsList, characterDetails); //*dev use
```
Let's remove console.log debug statements please.

We'll be deploying to Lambda and the way that logs make it into CloudWatch (AWS logging service), is via [stdout](https://en.wikipedia.org/wiki/Standard_streams). `console.log` sends data to stdout. Once applications start to mature and we have real production traffic, the quality of the logs captured becomes critically important.

I haven't created a standard and correct way to output logs in this project yet, but generally we will want a more organized approach than ad-hoc console.log statements.

That being said, console.log is really useful for development purposes, but just remember to remove them before PR's if you can.

briansimoni commented on src/lib/crawler/crawler.test.ts:
There's a bunch of issues prettier is complaining about in this file. If you feel like it would be easier, you can add a script command to package.json

`"prettier:write": "prettier --write src/**/*.ts"`

And then you could run

`npm run prettier:write`

<img width="1349" alt="image" src="https://user-images.githubusercontent.com/11894935/230724533-3d15b2b2-054b-406d-93c8-a3d41e360523.png">

briansimoni replied:
I personally have prettier just fix everything automatically each time I hit save.

briansimoni commented on src/lib/crawler/crawler.test.ts:
```diff
@@ -1,32 +1,133 @@
+    } else {
+      fail("No character_details found in the first matchDetails item");
+    }
+  }, 60000);
```
When it comes to unit tests, ideally we want them to run very fast. Longer tests increase build times, and make it more inconvenient to run them in general. The tests lose value the less we run them. A lot of testing fameworks i.e. Jest, have options like `--watch` where any time you make any change to any file, it instantly runs all tests for you. We throw that option away when we start adding super long running tests. Let's find a way to speed this up.

There is a time and a place for longer tests, but that should be a thoughtful and deliberate choice.

briansimoni commented on src/lib/crawler/crawler.ts:
```diff
@@ -1,68 +1,222 @@
-  teamname: string;
+}
+
+// MatchSummary & CharacterDetail merged
```
In my opinion, merged doesn't accurately describe this interface. Merging one interface with another sounds more like what happens when you `extend`

```TypeScript
interface MatchDetails extends CharacterDetail
```

Also, I haven't really done a style review yet, and it should really be more on me to set the style standards, but I think you should consider the usefulness of some of these comments. I can easily read the interface and see that an array CharacterDetail is a property of MatchDetails. See my suggestion for a comment that explains something that may not be so obvious.

```suggestion
/**
 * MatchDetails is an aggregate of all data that the crawler collects 
 * for a given matchId. These objects are intended to be stored
 * long term in some kind of database for further use and analysis
 */
```
Writing good comments and documentation is relatively difficult. In your studying, you might have seen tips where people aggressively encourage new developers to write comments, but a useless comment is worse than no comment in my opinion. My recommendation is try to think of the type of people who will be reading this code in the future including yourself.

briansimoni commented on src/lib/crawler/crawler.ts:
```diff
@@ -1,68 +1,222 @@
+    return response.data;
+  }
+
+  // responsible for fetching match details using the match IDs of each given matchSummaries array, combining the details to create a single object, and restricting concurrent operations to 32
```
This comment is actually more useful. I just don't like I have to scroll so far to read the whole thing.

briansimoni reviewed (changes requested):
I think there is a lot of opportunity to clean some things up (I mainly want them to run fast), but your code appears to work!

<img width="980" alt="image" src="https://user-images.githubusercontent.com/11894935/230726349-198770d7-ab7c-49ca-8332-bfc155007791.png">

<img width="751" alt="image" src="https://user-images.githubusercontent.com/11894935/230726377-ce18be14-f7a3-4c7d-8ad1-354bedcf5ad7.png">

briansimoni commented on src/lib/crawler/crawler.test.ts:
```diff
@@ -1,32 +1,133 @@
       "utf-8"
     );
+    const testJson = JSON.parse(
+      fs.readFileSync(path.resolve(__dirname, "test.json"), "utf-8")
```
It looks like you didn't commit this file, so tests are failing on my computer.

briansimoni commented on src/lib/crawler/crawler.test.ts:
```diff
@@ -1,32 +1,133 @@
+    // mocking fetchMatchData method to simulate delay
+    const originalFetchMatchData = WarmaneCrawler.prototype.fetchMatchData;
+    WarmaneCrawler.prototype.fetchMatchData = jest.fn(
+      async (matchId: string) => {
+        return new Promise((resolve) =>
+        setTimeout(() => resolve([] as CharacterDetail[]), 100)
+        );
+      }
```
This actually accomplishes the same thing that you're doing, but it doesn't introduce 100ms of latency each time it's called.

`async` implicitly returns a `Promise` so you shouldn't need to `new Promise()` here. Also, we're testing the inputs and outputs of functions. We don't need to simulate network latency with `setTimeout()`. At least, it shouldn't affect function output unless you have some kind of [race condition](https://en.wikipedia.org/wiki/Race_condition)
```suggestion
      async () => []
```

brennan-obc committed:
PR Review Corrections

##### Prettier Failure
- formatted with Prettier

##### Test Speed
- removed unnecessary timer
- lowered Jest test timer to 20s (tests will occasionally time out waiting on Warmane's response without) 
- all tests pass between 10 - 16 seconds

brennan-obc committed:
PR Review Corrections

- removed unnecessary comments, refined remainder
- added line breaks to all comments (no longer blinded by 'word-wrap: on')
- removed debug logs

brennan-obc committed:
Keeping test data in sync

- test.json and test.html will have data pulled from the same time; thus the same dataset

brennan-obc committed:
PR Review Corrections

- formatted w/ Prettier
- removed unnecessary timeout
crawler.test.ts
- changed test matchSummaries array length from 100 to 10, bringing average test duration from 17s to 4s
- removed Jest timeout argument from each function

crawler.ts
- removed unnecessary comments, refined remainder
- added line breaks to comments

test.html
- updated w/ data that will match test.json

test.json
- added test.json file

brennan-obc committed:
Merge branch 'v2' into v2.b

brennan-obc committed:
Fixed ESLint Issues

removed unused variables

briansimoni committed:
fix typescript error in src/routes.ts
